game/procurement.py

```diff
@@ -182,26 +182,49 @@ def most_needed_unit_class(self, cp: ControlPoint) -> GroundUnitClass:
     @staticmethod
     def fulfill_aircraft_request(
         squadrons: list[Squadron], quantity: int, budget: float
-    ) -> Tuple[float, bool]:
+    ) -> Tuple[float, bool, bool]:
+        """
+        Purchases aircraft for a requested task. Arguments:
+            List of potential squadrons to consider the task.
+            Requested quantity.
+            Available budget.
+        Returns:
+            A tuple containing, respectively, a float (remaining budget), a bool (true if the request was fulfilled)
+            and a bool (false if the request was not fulfilled because of insufficient pilots, prompting a retry).
+        """
         for squadron in squadrons:
+            insufficient_pilots = False
             price = squadron.aircraft.price * quantity
+
+            # Final check to make sure the number of aircraft won't exceed the number of available pilots
+            # after fulfilling this aircraft request.
+            if (
+                squadron.pilot_limits_enabled
+                and squadron.expected_size_next_turn + quantity > squadron.max_size
+            ):
+                insufficient_pilots = True
```

Comment:
For this to be correct it needs to happen _after_ the budget check. Right now this will return `True` for `insufficient_pilots` if _any_ squadron didn't have sufficient pilots. You only want to do that if _no_ squadron has enough pilots, otherwise you fall into the bad case described on L229.